File: apps/web/src/lib/api/ask.ts
import { createApiClient } from "./client";
import type { RetrieveMatch } from "./retrieve";

export type AskResponse = {
  question: string;
  answer: string | null;
  answer_status?: "retrieval_only" | "supported" | "insufficient_context" | null;
  answer_sources?: number[];
  source_id?: number | null;
  top_k: number;
  retrieval_only: boolean;
  matches: RetrieveMatch[];
};

export async function askQuestion(payload: {
  question: string;
  top_k?: number;
  source_id?: number;
  retrieval_only?: boolean;
}): Promise<AskResponse> {
  return createApiClient().post<AskResponse>("/api/ask", {
    question: payload.question,
    top_k: payload.top_k ?? 5,
    source_id: payload.source_id,
    retrieval_only: payload.retrieval_only ?? false,
  });
}

File: apps/api/app/schemas/ask.py
from typing import Literal

from pydantic import Field

from app.schemas.common import APIModel
from app.schemas.retrieve import RetrieveMatch


class AskRequest(APIModel):
    question: str = Field(min_length=1)
    top_k: int = Field(default=5, ge=1, le=10)
    source_id: int | None = None
    retrieval_only: bool = False


class AskResponse(APIModel):
    question: str
    answer: str | None
    answer_status: Literal["retrieval_only", "supported", "insufficient_context"] | None = None
    answer_sources: list[int]
    source_id: int | None = None
    top_k: int
    retrieval_only: bool
    matches: list[RetrieveMatch]
